python/lsst/verify/compatibility/metricsControllerTask.py

```diff
@@ -0,0 +1,420 @@
+            already contain a measurement for
+            ``metricTask.getOutputMetric()``.
+        metricTask : `lsst.verify.compatibility.MetricTask`
+            The metric to compute.
+        dataref : `lsst.daf.persistence.ButlerDataRef`
+            The repository and data ID to analyze. The data ID may be
+            incomplete, but must have the granularity of the desired metric.
+        """
+        self.log.debug("Running %s on %r", type(metricTask), dataref)
+        inputTypes = metricTask.getInputDatasetTypes(metricTask.config)
+        inputData = {}
+        inputDataIds = {}
+        for param, dataType in inputTypes.items():
+            inputRefs = dafPersist.searchDataRefs(
+                dataref.getButler(), dataType, dataId=dataref.dataId)
+            inputData[param] = [ref.get() for ref in inputRefs]
+            inputDataIds[param] = [ref.dataId for ref in inputRefs]
+
+        outputDataIds = {"measurement": dataref.dataId}
+        try:
+            result = metricTask.adaptArgsAndRun(inputData, inputDataIds,
+                                                outputDataIds)
+            value = result.measurement
+            if value is not None:
+                job.measurements.insert(value)
+        except MetricComputationError:
+            # Apparently lsst.log doesn't have built-in exception support?
+            self.log.error("Measurement of %r failed on %s->%s\n%s",
+                           metricTask, inputDataIds, outputDataIds,
+                           traceback.format_exc())
+
+    def runDataRefs(self, datarefs):
+        """Calls all registered metric tasks on each dataref.
+
+        This method shall load all datasets required to compute a particular
+        metric, and shall persist the metrics as one or more `lsst.verify.Job`
+        objects. Some objects may be persisted, or incorrectly persisted, in
+        the event of an exception.
+
+        Parameters
+        ----------
+        datarefs : `list` of `lsst.daf.persistence.ButlerDataRef`
+            The data to measure. Datarefs may be complete or partial; each
+            shall generate a measurement at the same granularity (e.g., a
+            dataref with only `"visit"` specified shall generate
+            visit-level measurements).
+
+        Returns
+        -------
+        struct : `lsst.pipe.base.Struct`
+            A Struct containing at least the following component:
+
+            - ``jobs`` : a `list` of collections of measurements
+              (`lsst.verify.Job`). Each job in the list shall contain the
+              measurement(s) for the corresponding dataref, and each job
+              shall have up to one measurement for each element
+              in `self.measurers`. A particular measurement shall be omitted
+              if it could not be created.
+        """
+        jobs = []
+        index = 0
+        for dataref in datarefs:
+            job = Job.load_metrics_package()
```

Comment:
I don't think you need to load the metrics package into the Job since you aren't working with `job.metrics` or `job.specs`. I think

```
job = Job()
```

is sufficient (and faster).

Reply:
Can you elaborate? While it would certainly makes testing easier for me, SQR-019 [recommends the current method].

Also, I noticed while working on a previous ticket that I get errors if I try to use the wrong units for a measurement value. Will those checks still happen (and if so, when/where) if I use a bare `Job`?

Reply:
The unit compatibility check will happen during `dispatch_verify`, which does run `Job.load_metrics_package` and integrates in the metric and specification with any new measurement data.

Unless you think tasks will need to do something based on metric and specification data, you can save time by not loading it at this point.